### PR TITLE
feat: implemented CVRP environment

### PR DESCRIPTION
**Pull request by surana01**
- Implemented CVRP environment
- Modified TSP - removed get_augmentations and small changes to docstrings
- Created utils file in 'jumanji/environments/combinatorial/ and added get_augmentations to it (also created test file for utils)
- Updated readme to include CVRP

### CODE REVIEW

**Review by clement-bonnet — DISMISSED** [review]
Great contribution, thank you!

**Review by clement-bonnet — APPROVED** [review]

File: jumanji/environments/combinatorial/cvrp/env.py [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+    generate_problem,
+)
+from jumanji.types import Action, TimeStep, restart, termination, transition
+
+
+class CVRP(Environment[State]):
+    """
+    Capacitated Vehicle Routing Problem (CVRP) environment as described in [1].
+    - observation: Observation
+        - coordinates: jax array (float32) of shape (num_nodes + 1, 2)
```
Comment (djbyrne) [inline jumanji/environments/combinatorial/cvrp/env.py:38]: docstrings are missing full stops

File: jumanji/environments/combinatorial/cvrp/env.py [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+    [1] Toth P., Vigo D. (2014). "Vehicle routing: problems, methods, and applications".
+    """
+
+    def __init__(
+        self, problem_size: int = 100, max_capacity: int = 30, max_demand: int = 10
+    ):
+        assert max_capacity >= max_demand, (
+            f"The demand associated with each node must be lower than the maximum capacity, "
+            f"hence the maximum capacity must be >= {max_demand}."
+        )
```
Comment (djbyrne) [inline jumanji/environments/combinatorial/cvrp/env.py:80]: this should be raised as an error instead of an assert. Asserts should only be used for debugging purposes as they can be turned off. Anything that is "user error" like creating an Env with capacity > demand should be an raised as an error :)

File: jumanji/environments/combinatorial/cvrp/env.py [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+        Run one timestep of the environment's dynamics.
+
+        Args:
+            state: State object containing the dynamics of the environment.
+            action: Array containing the index of the next node to visit.
+
+        Returns:
+            state, timestep: Tuple[State, TimeStep] containing the next state of the environment,
+            as well
+            as the timestep to be observed.
```
Comment (djbyrne) [inline jumanji/environments/combinatorial/cvrp/env.py:132]:
```suggestion
            as well as the timestep to be observed.
```

File: jumanji/environments/combinatorial/cvrp/specs.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+        """
+        all_kwargs = {
+            "coordinates_spec": self.coordinates_spec,
+            "demands_spec": self.demands_spec,
+            "position_spec": self.position_spec,
+            "capacity_spec": self.capacity_spec,
+            "action_mask_spec": self.action_mask_spec,
+        }
+        all_kwargs.update(kwargs)
+        return ObservationSpec(**all_kwargs)  # type: ignore
```
Comment (djbyrne) [inline jumanji/environments/combinatorial/cvrp/specs.py:110]:
```suggestion
        all_kwargs.update(kwargs)
        
        return ObservationSpec(**all_kwargs)  # type: ignore
```

File: jumanji/environments/combinatorial/cvrp/specs_test.py [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+import pytest
+
+from jumanji.environments.combinatorial.cvrp.env import CVRP
+from jumanji.environments.combinatorial.cvrp.types import Observation
+
+
+class TestObservationSpec:
+    env = CVRP()
+    observation_spec = env.observation_spec()
+    observation = observation_spec.generate_value()
```
Comment (djbyrne) [inline jumanji/environments/combinatorial/cvrp/specs_test.py:28]: We should avoid defining variables at the class level as we add in the possibility that they get changed during testing. Instead we should put them in a setup() that is called for each test or we put them in as fixtures. Happy to go with either. My preference is fixtures

File: jumanji/environments/combinatorial/cvrp/specs_test.py [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+
+    def test_observation_spec__generate_value(self) -> None:
+        """Test generating a value which conforms to the observation spec."""
+        assert isinstance(self.observation, Observation)
```
Comment (djbyrne) [inline jumanji/environments/combinatorial/cvrp/specs_test.py:32]: Would prefer this to take in the env as a fixture or as an env setup before the test, then call env.observation as an instance. This ensures nothing has changed since 'observation' was initially created.

File: jumanji/environments/combinatorial/cvrp/specs_test.py [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+        # Check that a different dtype breaks the validation
+        modified_dtype_observation = observation._replace(
+            coordinates=observation.coordinates.astype(jnp.float16)
+        )
+        with pytest.raises(ValueError):
+            self.observation_spec.validate(modified_dtype_observation)
+
+        # Check that validating another object breaks the validation
+        with pytest.raises(Exception):
+            self.observation_spec.validate(None)  # type: ignore
```
Comment (djbyrne) [inline jumanji/environments/combinatorial/cvrp/specs_test.py:53]: Ideally we should have each test assert 1 thing. These should probably be split into 3 separate tests, each testing an edge case of validate.

File: jumanji/environments/combinatorial/cvrp/specs_test.py [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+            ("demands_spec", observation_spec.demands_spec.replace(shape=(3, 4))),
+            ("position_spec", observation_spec.position_spec.replace(name="new_name")),
+            ("capacity_spec", observation_spec.capacity_spec.replace(name="new_name")),
+            (
+                "action_mask_spec",
+                observation_spec.action_mask_spec.replace(shape=(3, 4)),
+            ),
+        ],
+    )
+    def test_observation_spec__replace(self, arg_name: str, new_value: Any) -> None:
```
Comment (djbyrne) [inline jumanji/environments/combinatorial/cvrp/specs_test.py:71]: Can we add a short description of what this is testing and what we expect to happen

File: jumanji/environments/combinatorial/cvrp/specs_test.py [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+        old_spec = self.observation_spec
```
Comment (djbyrne) [inline jumanji/environments/combinatorial/cvrp/specs_test.py:72]: This should be fine, but would like this to be taking in observation_spec as a fixture instead of taking the value of it

File: jumanji/environments/combinatorial/cvrp/specs_test.py [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+        new_spec = old_spec.replace(**{arg_name: new_value})
+        assert new_spec != old_spec
+        assert getattr(new_spec, arg_name) == new_value
+        for attr_name in {
+            "coordinates_spec",
+            "demands_spec",
+            "position_spec",
+            "capacity_spec",
+            "action_mask_spec",
+        }.difference([arg_name]):
```
Comment (djbyrne) [inline jumanji/environments/combinatorial/cvrp/specs_test.py:82]:
```suggestion
        arg_names = {
            "coordinates_spec",
            "demands_spec",
            "position_spec",
            "capacity_spec",
            "action_mask_spec",
        }.difference([arg_name])
        
        for attr_name in arg_names:
```
Could we seperate out defining the gaurd and the loop itself, as the gaurd is quite big. Having it in one line is a bit difficult to read. (obviously this can be done cleaner than what I have, but just to make it clear what I mean :wink:   )

File: jumanji/environments/combinatorial/utils_test.py [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+from jax import numpy as jnp
+
+from jumanji.environments.combinatorial.utils import get_coordinates_augmentations
+
+
+def test_utils__get_coordinates_augmentations() -> None:
+    """Checks that the augmentations of a given instance problem is computed properly."""
+    coordinates = jnp.array([[0.65, 0.85], [0.18, 0.06], [0.41, 0.19], [0.92, 0.27]])
+
+    expected_augmentations = jnp.array(
```
Comment (djbyrne) [inline jumanji/environments/combinatorial/utils_test.py:25]: Can we make coordindates and expected_augmentations fixtures? I know they are only used in this test, but IMO it will be easier to read and expand these tests

**Review by djbyrne — COMMENTED** [review]
Looks really good! some very minor comments around testing, but other than that LGTM :+1:

**Review by clement-bonnet — APPROVED** [review]
